CI: refactor TeamCity build configurations.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR proposes a major refactor of the build configurations for `Calypso` and `WPCOM` projects in TeamCity.

Key changes:
- new methods that implement E2E-related steps.
- new methods that return predefined values for artifact rules, VCS settings and more.

Details:
The current configuration of WPCOM and Calypso build configurations is that of chaos and discrepancy.
There exist numerous similar copies of configuration or build steps that are scattered across these two files.

The goal here is two fold:
1. unify the configuration and build steps where possible, and;
2. eliminate potential mistakes that may arise in the future when changes spanning every build configuration needs to be made.

For 1., some build configurations did not have the updated artifact rules applied, which meant trace files were being ignored. Similarly, some build configurations were missing certain environment variables. 

For 2., by extracting the common parts out to methods the DRY design principle is applied to make future changes easier and less error-prone.

#### Testing instructions

All affected builds under `Calypso` and `WPCOM` projects must be able to:
- run a build.
- launch the `run e2e` step.
- upload artifacts (if any).
- report appropriate failure status.

Build configurations:
- Calypso
  - [x] mobile
  - [x] desktop
  - [x] pre-release
  - [x] quarantined
- WPCOM
  - [x] mobile
  - [x] desktop
  - [x] mobile coblocks
  - [x] desktop coblocks
  - [x] i18n
  - [x] p2

Closes https://github.com/Automattic/wp-calypso/issues/59249.

worldomonation committed:
Rename `prepareEnvironment.kt` to `Steps.kt`.
Add new function for step `collectResults`.

worldomonation commented on .teamcity/_self/lib/e2e/Steps.kt:
```diff
@@ -0,0 +1,179 @@
+ */
+fun BuildSteps.runTests(
+	stepName: String = "",
+	dockerBuildNumber: String = "",
```
This is used by the Playwright E2E (mobile/desktop) variants.

worldomonation commented on .teamcity/_self/lib/e2e/Steps.kt:
```diff
@@ -0,0 +1,179 @@
+					echo ${'$'}URL
+					exit 1
+				fi
+
```
These empty spaces were originally left to make the final output of `scriptContentBuilder.toString()` look nicer, but it isn't actually printed anywhere. Perhaps it should be removed.

worldomonation commented on .teamcity/_self/lib/e2e/Steps.kt:
```diff
@@ -0,0 +1,179 @@
+		)
+
+		for ((key, value) in envVars) {
+			scriptContentBuilder.appendln( "export $key=$value\n".trimIndent())
```
This is where the envVars map is exported as individual environment variable values.

scinos replied:
The point of having to export each key explicitly was to facilitate searching where each key is used and provide an easier way to copy/paste to run the tests locally.

If we are going to export all of them automatically, we can as well just use TeamCity's mechanism for exporting them (i.e. use `%env.KEY%`)

worldomonation replied:
Gotcha. I wasn't sure of the reason for the explicit `export`s, but what you said makes sense. 

I'll move to using a `env.KEY` type declaration instead.

worldomonation commented on .teamcity/_self/lib/e2e/VCS.kt:
```diff
@@ -0,0 +1,13 @@
+/**
+ *
+ */
+fun BuildType.wpCalypsoVCS( enforceCleanCheckout: Boolean = true ): Unit {
```
I'm open to a better function name.

worldomonation commented:
All build configurations that are affected are able to run as expected.

Many build configurations within WPCOM are permanently failing and so their failures are not related to this change.

scinos commented on .teamcity/_self/lib/e2e/Steps.kt:
```diff
@@ -0,0 +1,179 @@
+fun BuildSteps.prepareEnvironment(): ScriptBuildStep {
+	return script {
+		name = "Prepare environment"
+		scriptContent = """
```
Can we use `bashNodeScript` instead of repeating bash options and the command to update node?

worldomonation replied:
Now using `bashNodeScript`.

scinos commented on .teamcity/_self/lib/e2e/Steps.kt:
```diff
@@ -0,0 +1,179 @@
+ *	- Logs
+ *	- Trace files
+ */
+fun BuildSteps.collectResults(): ScriptBuildStep {
```
It would make sense to move this function to `.teamcity/_self/lib/e2e/ArtifactRules.kt`, as both are related to artifacts.

worldomonation replied:
I'd say that all steps should be kept together. 

However I'll keep your suggestion in mind as I re-evaluate the current approach I've taken as per your comment in https://github.com/Automattic/wp-calypso/pull/59465#discussion_r773593990.

worldomonation replied:
I've shifted to use a class (similar to `WPComPluginBuild`) and so `ArtifactRules.kt` has been removed.

scinos commented on .teamcity/_self/lib/e2e/Steps.kt:
```diff
@@ -0,0 +1,179 @@
+			export TEAMCITY_VERSION=2021
+			export DEBUG=pw:api
+			export HEADLESS=false
+			export LOCALE=en
```
This var doesn't seem specifically related to TeamCity. Maybe this should be the default value so we don't need to specify it.

worldomonation replied:
Changed to set these values as default.

scinos commented on .teamcity/_self/projects/WPComTests.kt:
```diff
@@ -219,16 +221,9 @@ fun gutenbergPlaywrightBuildType( targetDevice: String, buildUuid: String ): Bui
-			screenshots => screenshots
-			trace => trace
-		""".trimIndent()
+		artifactRules = artifactRules()
```
What if we create an extension of `BuildType` (for example, `e2eBuildType`) that already has common parts (vcs, artifacts, features) already assigned? It doesn't feel very OOP-y to create an object `BuildType` and always assign the same value to the same parameters.

worldomonation replied:
That's definitely an idea...I'll take a stab at it. With the time we have left counting down I want to make sure I get this done before you leave.

worldomonation replied:
I created a custom build type as follows:

```
fun calypsoE2EBuildType(): BuildType {
  // build type definitions
  steps {
      prepareEnvironment()
      collectResults()
  }
}
```

Note the partially-populated `steps`. I then want to call this from a build configuration:

```
var config = calypsoE2EBuildType()

config.steps.items.add(1,  runTests( <params> ) )
```

In the above, I want to insert the `runTests` step in between the `prepareEnvironment()` and `collectResults` steps, but this does not appear to work. Do you have any ideas?

worldomonation replied:
Never mind, I used the `WPComPluginBuild` class as reference - my mistake was trying to use a `fun` for that purpose.

Now I am blocked by an odd issue where if I use the following combination:

Playwright E2E build configuration
-> calls runTests
    -> calls bashNodeScript

the dependencies are not being recognized as having been installed on the system, despite it being so:

```
   browserType.launch: Host system is missing dependencies!

11:59:36   

11:59:36     Missing libraries are:

11:59:36         libnss3.so

11:59:36         libnssutil3.so

11:59:36         libsmime3.so

11:59:36         libnspr4.so

11:59:36         libatk-1.0.so.0

11:59:36         libatk-bridge-2.0.so.0

11:59:36         libcups.so.2

11:59:36         libdrm.so.2

11:59:36         libdbus-1.so.3

11:59:36         libxkbcommon.so.0

11:59:36         libXcomposite.so.1

11:59:36         libXdamage.so.1

11:59:36         libXfixes.so.3

11:59:36         libXrandr.so.2

11:59:36         libgbm.so.1

11:59:36         libasound.so.2

11:59:36         libatspi.so.0
```

(build https://teamcity.a8c.com/buildConfiguration/calypso_Calypso_E2E_Playwright_mobile/7175344?buildTab=log&focusLine=484&expandAll=true)

This issue goes away if I change the call structure to:

Playwright E2E build configuration
-> calls bashNodeScript

The build configuration is able to execute the tests as expected.

(build https://teamcity.a8c.com/buildConfiguration/calypso_Calypso_E2E_Playwright_mobile/7175410)

Rather confusing as to why this is the case, but I'm still investigating.

worldomonation replied:
I think I've got it.

It was due to the missing line `dockerImage = "%docker_image_e2e%"`.

worldomonation replied:
Resolved. Now using `class E2EBuildType` which sets these values as default.

worldomonation commented:
Sorry for the repeated notifications. I can't seem to figure out how to add a step after calling the custom build type.

worldomonation committed:
Use bashNodeScript directly in calypsoE2EBuildType and fill in the steps using other parameters.

worldomonation committed:
Specify docker image

worldomonation committed:
Migrate all build configurations

worldomonation committed:
Rename class template to E2EBuildType

worldomonation commented on .teamcity/_self/lib/customBuildType/E2EBuildType.kt:
```diff
@@ -0,0 +1,144 @@
+import jetbrains.buildServer.configs.kotlin.v2019_2.triggers.schedule
+import jetbrains.buildServer.configs.kotlin.v2019_2.triggers.vcs
+
+open class E2EBuildType(
```
I've followed the template set by `WPComPluginBuild` for this class.

Define an instance like:

```kotlin
private object xyz : E2EBuildType(
  // required params
)
```

worldomonation committed:
Added JSDoc and comments

worldomonation commented on .teamcity/_self/lib/customBuildType/E2EBuildType.kt:
```diff
@@ -0,0 +1,170 @@
+						token = "credentialsJSON:57e22787-e451-48ed-9fea-b9bf30775b36"
+					}
+				}
+			}
```
For simplicity's sake, I have chosen to enable this feature by default in the class. 

Previously, it looks like not all build configurations had this feature enabled (eg. `QuarantinedE2ETests`). If I understand this feature correctly, having it enabled by default shouldn't have an impact.

worldomonation commented on .teamcity/_self/lib/customBuildType/E2EBuildType.kt:
```diff
@@ -0,0 +1,170 @@
+
+		dependencies.buildDependencies()
+
+		failureConditions {
+			executionTimeoutMin = 20
+			// Don't fail if the runner exists with a non zero code. This allows a build to pass if the failed tests have been muted previously.
+			nonZeroExitCode = false
+
+			// Fail if the number of passing tests is 50% or less than the last build. This will catch the case where the test runner crashes and no tests are run.
+			failOnMetricChange {
+				metric = BuildFailureOnMetric.MetricType.PASSED_TEST_COUNT
+				threshold = 50
+				units = BuildFailureOnMetric.MetricUnit.PERCENTS
+				comparison = BuildFailureOnMetric.MetricComparison.LESS
+				compareTo = build {
+					buildRule = lastSuccessful()
+				}
+			}
+		}
```
Some builds previously only had the `executionTimeoutMin` and `nonZeroExitCode` failure conditions, but in a similar story as above I've chosen to include `failOnMetricChange` as another default failure condition.

noahtallen commented on .teamcity/_self/projects/WPComTests.kt:
```diff
@@ -212,102 +213,40 @@ fun gutenbergBuildType(screenSize: String, buildUuid: String): BuildType {
+		buildParams = {
 			text(
-				name = "URL",
+				name = "env.URL",
```
Does this need to use `"%env.URL%"`? I seem to remember having to use that to access params in the past.

worldomonation replied:
It's been a few weeks so it's a bit rusty, but what I recall is that TeamCity configuration params do not get propagated to the builds, so when we go to set the environment variable to %URL% it tells me that the value is undefined.

noahtallen reviewed (commented):
Looking pretty good! I love how much duplication we're removing here :D

worldomonation commented:
@noahtallen Re-requesting review - for the comment you had in https://github.com/Automattic/wp-calypso/pull/59465#discussion_r777733152 I left a response but no changes were made.

worldomonation committed:
Rebase onto newest trunk 92def8d0e89f008a108b87ee63749fee9bea12f8 and incorporate changes:
- remove auto-trigger for Calypso E2E (mobile/desktop) at 0/30 past the hour on trunk.

worldomonation committed:
Switch over to using E2E_CONFIG_ENCRYPTION_KEY.

worldomonation committed:
Rebase artifact

worldomonation commented on .teamcity/_self/lib/customBuildType/E2EBuildType.kt:
```diff
@@ -0,0 +1,170 @@
+	var buildUuid: String,
+	var buildName: String,
+	var buildDescription: String,
+	var concurrentBuilds: Int = 0,
```
The only user of this param is the Pre-Release Tests, which has a maximum concurrent build count of 1.

worldomonation commented on .teamcity/_self/lib/customBuildType/E2EBuildType.kt:
```diff
@@ -0,0 +1,170 @@
+	var buildName: String,
+	var buildDescription: String,
+	var concurrentBuilds: Int = 0,
+	var getCalypsoLiveURL: String = "",
```
The only consumer of this param is the CalypsoE2E build type which uses live container to run its tests.

worldomonation commented on .teamcity/_self/lib/customBuildType/E2EBuildType.kt:
```diff
@@ -0,0 +1,170 @@
+		}
+
+		params {
+			param("env.NODE_CONFIG_ENV", "test")
+			param("env.PLAYWRIGHT_BROWSERS_PATH", "0")
+			param("env.TEAMCITY_VERSION", "2021")
+			param("env.HEADLESS", "false")
+			param("env.LOCALE", "en")
+			param("env.DEBUG", "pw:api")
+			buildParams()
```
I've chosen in this refactor to use environment variables by default for all parameters.

Though several weeks has elapsed, I recall the couple of factors that led to this:
- references to configuration parameters (unprefixed params) were failing, with TeamCity stating that it was undefined.
- may (if not all) of the parameters defined here end up being exported as an environment variable later in the script.

noahtallen commented on .teamcity/_self/lib/customBuildType/E2EBuildType.kt:
```diff
@@ -0,0 +1,170 @@
+					# environment variable within this step - therefore the export
+					# for NODE_CONFIG variable has to be done here instead of
+					# within the Kotlin DSL.
+					export NODE_CONFIG="{\"calypsoBaseURL\":\"${'$'}{URL%/}\"}"
```
Gotta love all the escaping :sweat_smile: 

So this comes down to: `"calypsoBaseURL": "${URL%}"}`. Out of curiosity, what does the `%` do here?

worldomonation replied:
That's a good question - I took a look at the current config, and the extraneous `%` is present. This suggests to me that it was left over from a time when the `%URL%` configuration parameter was used.

I'll remove the extraneous %.

worldomonation replied:
I tried to further simplify how `URL` param was set and it _looked like_ it was successful on first glance.

Unfortunately I had to revert the attempts I made because the successful Calypso E2E runs were due to it being run against the standard `calypsoBaseURL: https://wordpress.com` instead of the live branches.

Until we have a better way to obtain the container name, this weird handling of URL/NODE_CONFIG variable will have to stay.

noahtallen commented on .teamcity/_self/lib/customBuildType/E2EBuildType.kt:
```diff
@@ -0,0 +1,170 @@
+			perfmon {
+			}
+			commitStatusPublisher {
+				vcsRootExtId = "${Settings.WpCalypso.id}"
```
Unimportant, but it might not need to be inside a string. Maybe it could just be `vcsRootExtId = Settings.WpCalypso.id`

worldomonation replied:
I think so too - this was just copy/pasted over from the original config. Thanks!

worldomonation replied:
I tried removing the string part, but it seems it is required, otherwise a compilation error results:

```
Compilation error _self/lib/customBuildType/E2EBuildType.kt[156:20]: Type mismatch: inferred type is Id? but String? was expected
```

noahtallen replied:
Unfortunate!

noahtallen commented on .teamcity/_self/lib/e2e/Steps.kt:
```diff
@@ -44,3 +46,41 @@ fun BuildSteps.prepareEnvironment(): ScriptBuildStep {
+ *	- Logs
+ *	- Trace files
+ */
+fun BuildSteps.collectResults(): ScriptBuildStep {
```
You might be able to move this back into the body of the main class by using `bashNodeScript`. I think it sets nearly all of the script properties in the same way

worldomonation replied:
Sure. I had kept the `collectResults` step split out in case other classes may want to use it, but that doesn't appear to be the case for now.

worldomonation replied:
I ended up moving both `collectResults` and `prepareEnvironment` out to be defined in the class instead.

noahtallen reviewed (approved):
This is an incredible improvement, thank you! Just a few tiny comments, but overall I think this is good to go.

We should wait to merge until we have a full day ahead to fix anything, though, since it's such a big change!

worldomonation committed:
- move `commitStatusPublisher.vcsRootExtId` out of string.
- move `collectResults` step into the class.
- move `prepareEnvironment` step into the class.

worldomonation committed:
Fix errors
- remove extrneous character
- import jetbrains.buildServer.configs.kotlin.v2019_2.BuildStep
- restore string for Settings.WpCalypso.id

worldomonation committed:
Add comments why the `getCalypsoLiveURL` exists

worldomonation committed:
Remove trailing /

worldomonation committed:
Fix DataHelper.adjustInviteLink's handling of URL:
- URL for livebranch build configurations have a trailing '/' which causes issues with the way `DataHelper.adjustInviteLink` was (naively) handling the URL.
